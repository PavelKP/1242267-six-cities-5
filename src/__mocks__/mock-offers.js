export const offers = [
  {
    "city": {
      "name": `Cologne`,
      "coordinates": [
        50.938361,
        6.959974
      ],
      "zoom": 13
    },
    "images": [
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/1.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/9.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/2.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/5.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/7.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/6.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/15.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/4.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/18.jpg`
    ],
    "title": `Nice, cozy, warm big bed apartment`,
    "rating": 2.5,
    "bedrooms": 2,
    "price": 485,
    "host": {
      "avatar": `img/avatar-angelina.jpg`,
      "id": 25,
      "isPro": true,
      "name": `Angelina`
    },
    "description": [
      `Peaceful studio in the most wanted area in town. Quiet house Near of everything. Completely renovated. Lovely neighbourhood, lot of trendy shops, restaurants and bars in a walking distance.`
    ],
    "id": 1,
    "isBookmarked": false,
    "isPremium": false,
    "entire": `house`,
    "adults": 6,
    "inside": [
      `Breakfast`,
      `Washer`,
      `Baby seat`,
      `Laptop friendly workspace`,
      `Air conditioning`
    ],
    "coordinates": [
      50.945361,
      6.935974
    ],
    "zoom": 16,
    "preview": `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`
  }
];

export const rawOffers = [
  {
    "city": {
      "name": `Cologne`,
      "location": {
        "latitude": 50.938361,
        "longitude": 6.959974,
        "zoom": 13
      }
    },
    "preview_image": `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`,
    "images": [
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/1.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/9.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/2.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/5.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/7.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/6.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/15.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/4.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/18.jpg`
    ],
    "title": `Nice, cozy, warm big bed apartment`,
    "is_favorite": false,
    "is_premium": false,
    "rating": 2.5,
    "type": `house`,
    "bedrooms": 2,
    "max_adults": 6,
    "price": 485,
    "goods": [
      `Breakfast`,
      `Washer`,
      `Baby seat`,
      `Laptop friendly workspace`,
      `Air conditioning`
    ],
    "host": {
      "id": 25,
      "name": `Angelina`,
      "is_pro": true,
      "avatar_url": `img/avatar-angelina.jpg`
    },
    "description": `Peaceful studio in the most wanted area in town. Quiet house Near of everything. Completely renovated. Lovely neighbourhood, lot of trendy shops, restaurants and bars in a walking distance.`,
    "location": {
      "latitude": 50.945361,
      "longitude": 6.935974,
      "zoom": 16
    },
    "id": 1
  }
];
