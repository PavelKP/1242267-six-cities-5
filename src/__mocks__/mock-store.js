
const DEFAULT_SORTING = `popular`;

export const AuthorizationStatus = {
  NO_AUTH: `NO_AUTH`,
  AUTH: `AUTH`,
};

const city = {
  name: `Moscow`,
  coordinates: [111, 222],
  zoom: 12,
};

const cities = [city];

const offers = [
  {
    city: {
      name: `Moscow`,
      coordinates: [
        50.938361,
        6.959974
      ],
      zoom: 13
    },
    images: [
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/15.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/9.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/4.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/19.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/2.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/11.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/1.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/17.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/5.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`
    ],
    title: `The house among olive `,
    rating: 3.8,
    bedrooms: 3,
    price: 493,
    host: {
      avatar: `img/avatar-angelina.jpg`,
      id: 25,
      isPro: true,
      name: `Angelina`,
    },
    description: [
      `Discover daily local life in city center, friendly neighborhood, clandestine casino, karaoke, old-style artisans, art gallery and artist studio downstairs.`
    ],
    id: 1,
    isBookmarked: true,
    isPremium: false,
    entire: `house`,
    adults: 4,
    inside: [
      `Laptop friendly workspace`,
      `Washer`,
      `Towels`,
      `Fridge`,
      `Air conditioning`,
      `Baby seat`,
      `Breakfast`
    ],
    coordinates: [
      50.957361,
      6.9509739999999995
    ],
    zoom: 16,
    preview: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg`
  }
];

export default {
  DATA: {
    offers,
    cities,
  },
  INTERFACE: {
    city,
    activeSorting: DEFAULT_SORTING,
  },
  USER: {
    authorizationStatus: AuthorizationStatus.NO_AUTH,
    userData: {
      avatarUrl: `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/6.jpg`,
      email: `test@test.ru`,
    }
  }
};
