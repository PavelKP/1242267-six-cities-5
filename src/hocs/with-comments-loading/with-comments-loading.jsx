import React from 'react';
import {fetchCurrentReview} from '../../store/api-actions';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {reviewsPropTypes} from '../../prop-types/prop-types';

const withCommentsLoading = (Component) => {
  class WithCommentsLoading extends React.PureComponent {
    constructor(props) {
      super(props);

      this.state = {
        loading: true
      };

      this._offerId = this.props.offerId;
    }

    componentDidMount() {
      this.props.loadReview(this._offerId).then(() => {
        this.setState((state) => ({loading: !state.loading}));
      });
    }

    render() {
      // rest is the best, but eslint config doesn't allow this
      // const {loadReview, offerId, ...rest} = this.props;
      // Remove unnecessary props
      const {reviews} = this.props;

      return <Component loading={this.state.loading} reviews={reviews} />;
    }
  } // End of class

  WithCommentsLoading.propTypes = {
    offerId: PropTypes.number.isRequired,
    reviews: reviewsPropTypes.reviews,
    loadReview: PropTypes.func.isRequired,
  };

  const mapStateToProps = ({DATA}) => ({
    reviews: DATA.reviews,
  });

  const mapDispatchToProps = (dispatch) => ({
    loadReview(id) {
      return dispatch(fetchCurrentReview(id));
    }
  });

  // End of function - connect is HOC and returns wrapped WithCommentsLoading
  return connect(mapStateToProps, mapDispatchToProps)(WithCommentsLoading);
};


/*
  SECOND SOLUTION - redux compose feature

  https://ru.reactjs.org/docs/higher-order-components.html
  https://medium.com/practo-engineering/connected-higher-order-component-hoc-93ee63c91526

  import {compose} from 'redux';

  ...

const mapStateToProps = ({DATA}) => ({
  reviews: DATA.reviews,
});

const mapDispatchToProps = (dispatch) => ({
  loadReview(id) {
    return dispatch(fetchCurrentReview(id));
  }
})

const composedWithCommentsLoading = compose(
    connect(mapStateToProps, mapDispatchToProps),
    withCommentsLoading
);

export {withCommentsLoading};
export default composedWithCommentsLoading;
*/

export {withCommentsLoading};
export default withCommentsLoading;
