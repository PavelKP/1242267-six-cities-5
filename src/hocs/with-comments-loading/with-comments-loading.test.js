import React from 'react';
import renderer from 'react-test-renderer';
import PropTypes from 'prop-types';
import {withCommentsLoading} from './with-comments-loading';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import {offers} from '../../__mocks__/mock-offers';
import mockStoreData, {reviews} from '../../__mocks__/mock-store';
import {Router} from 'react-router-dom';
import browserHistory from '../../browser-history';
import thunk from 'redux-thunk';

const api = {
  get: () => Promise.resolve({
    data: reviews,
  })
};

const mockStore = configureStore([thunk.withExtraArgument(api)]);

const MockComponent = (props) => {
  const {children} = props;

  return (
    <div>
      {children}
    </div>
  );
};

const MockComponentWrapped = withCommentsLoading(MockComponent);

let store;
beforeEach(() => {
  store = mockStore(mockStoreData);
});

it(`withCommentsLoading is rendered correctly`, () => {

  const tree = renderer.create(
      <Provider store={store}>
        <Router history={browserHistory}>
          <MockComponentWrapped
            offers={offers}
            offerId={1}
            reviews={reviews}
            loadReview={()=>{}}
            loading={false}
          />
        </Router>
      </Provider>)
      .toJSON();

  expect(tree).toMatchSnapshot();
});

MockComponent.propTypes = {
  children: PropTypes.node,
};
