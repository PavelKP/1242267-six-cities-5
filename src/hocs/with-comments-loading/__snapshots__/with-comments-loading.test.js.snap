// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withCommentsLoading is rendered correctly 1`] = `<div />`;
